### ISSUE TITLE
Using cpp generated pb within python

### ISSUE THREAD

**Issue by IvanHristovVia** [protocolbuffers/protobuf#9903]
## 🐞 Issue

As can be seen from the [benchmarks](https://github.com/protocolbuffers/protobuf/blob/main/docs/performance.md), `python pb` show an immense performance boost while using the option with `c++` generated code. I tried searching for docs on how to use it but found none except a [thread](https://groups.google.com/g/protobuf/c/z7E80KYJscc?pli=1) dating back to 2010. I then tried the [solution](https://groups.google.com/g/protobuf/c/z7E80KYJscc/m/ysCjHHmoraUJ) but no significant change in performance occurred.

## Details

In more details what I did was to (follow the solution):

1. Create an empty `my_module.cc` module that imports `Python.h` as can be seen from the solution.
1. Generate `cpp` code from the `protobuf` definitions and move it to the same directory as `my_module.c`.
1. Import the generated `cpp` code header from within `my_module.cc`.
1. Add an extension to the `setup.py` file and build a shared object for `my_module`.
1. Import `my_module` within the `python` code before importing any other modules.
1. Benchmark the parsing performance of the code and observe no significant changes.

## ❓ Questions

1. Is the option is still relevant and viable?
1. Is there any proper guide on how to implement it? (not just some long forgotten thread)
1. Is there a way to check if I'm actually using the option, i.e. whether the message creation process is delegated to the `cpp DescriptorPool/MessageFactory`?

Thank you in advance.

---
PS - This is different from using only the recommended option with the `c` extension of `protobuf`. As I understand it when using `cpp` generated code the `c` extension should switch to directly using the `cpp` code instead of reflections of `python` code.

**Comment by haberman** [comment]
We are on the cusp of releasing a new version of Python that will give a large performance boost *without* requiring you to compile any custom C++ protos into your module.

We have announced the new release here: https://docs.google.com/document/d/12uHZ3-UFiXjVDBwbyy2qIATOdALhZY-asLILoyiYTQk/edit?resourcekey=0--0lRbPeIPBQ_I9Xc5oU2ZA

Please try out 4.21.0 (a release candidate should be posted to PyPI this week). It should completely resolve this problem.

**Comment by IvanHristovVia** [comment]
Thank you @haberman!  Did you mean version 3.21.0 (instead of 4) because we're at 3.20.1 now? Or am I getting it wrong. 😄

PS - Sounds great. A stupid question but I still have to use the c-extension, right?

**Comment by haberman** [comment]
> Thank you @haberman! Did you mean version 3.21.0 (instead of 4) because we're at 3.20.1 now?

The new release will be a new major version of Python, because there are some minor breaking changes in Python. However we're keeping the minor version synced to all the other languages so that we can easily know which versions were released together. For details see: https://developers.google.com/protocol-buffers/docs/news/2022-05-06#versioning

(btw I just realized I pasted the wrong link above, and I've corrected it. The Python announcement is here: https://developers.google.com/protocol-buffers/docs/news/2022-05-06#python-updates

> PS - Sounds great. A stupid question but I still have to use the c-extension, right?

If you want good performance, you'll want to use the C extension. But we're distributing binaries wheels for all major platforms, so you shouldn't have to compile any C code yourself. Does that answer your question?
